File: app/src/main/java/com/example/theweatheriswhat/CitiesWeatherActivity.java
package com.example.theweatheriswhat;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import android.os.Bundle;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

public class CitiesWeatherActivity extends AppCompatActivity {

    TabLayout tlWeather;
    ViewPager2 vpWeather;
    ListCitiesFragment listCitiesFragment;
    WeatherDetailsFragment weatherDetailsFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cities_weather);

        tlWeather = findViewById(R.id.tlWeather);
        vpWeather = findViewById(R.id.vpWeather);

        listCitiesFragment = new ListCitiesFragment();
        weatherDetailsFragment = new WeatherDetailsFragment();

        WeatherFragmentAdapter myWeathersFragmentsAdapter = new WeatherFragmentAdapter(this);
        vpWeather.setAdapter(myWeathersFragmentsAdapter);
        vpWeather.setOffscreenPageLimit(1);

        new TabLayoutMediator(tlWeather, vpWeather, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                if (position == 0) {
                    tab.setText("Cities List");
                } else if (position == 1) {
                    tab.setText("Weather Detail");
                }
            }
        }).attach();

    }

    public void getClickedCityName(String cityName) {
        vpWeather.setCurrentItem(1);
        weatherDetailsFragment.getCityNameFromActivityAndRefreshWeatherDetails(cityName);
    }

    public class WeatherFragmentAdapter extends FragmentStateAdapter {

        public WeatherFragmentAdapter(@NonNull FragmentActivity fragmentActivity) {
            super(fragmentActivity);
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            if (position == 0) {
                return listCitiesFragment;
            } else if (position == 1) {
                return weatherDetailsFragment;
            }
            return null;
        }

        @Override
        public int getItemCount() {
            return 2;
        }
    }
}

File: app/src/main/java/com/example/theweatheriswhat/CitiesWeatherActivity.kt
package com.example.theweatheriswhat

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

class CitiesWeatherActivity : AppCompatActivity() {
    private lateinit var tlWeather: TabLayout
    private lateinit var vpWeather: ViewPager2
    lateinit var listCitiesFragment: ListCitiesFragment
    lateinit var weatherDetailsFragment: WeatherDetailsFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cities_weather)
        tlWeather = findViewById(R.id.tlWeather)
        vpWeather = findViewById(R.id.vpWeather)
        listCitiesFragment = ListCitiesFragment()
        weatherDetailsFragment = WeatherDetailsFragment()
        val myWeathersFragmentsAdapter = WeatherFragmentAdapter(this)
        vpWeather.adapter = myWeathersFragmentsAdapter
        vpWeather.offscreenPageLimit = 1
        TabLayoutMediator(tlWeather, vpWeather) { tab, position ->
            if (position == 0) {
                tab.text = "Cities List"
            } else if (position == 1) {
                tab.text = "Weather Detail"
            }
        }.attach()
    }

    fun getClickedCityName(cityName: String?) {
        vpWeather.currentItem = 1
        cityName?.let { weatherDetailsFragment.getCityNameFromActivityAndRefreshWeatherDetails(it) }
    }

    inner class WeatherFragmentAdapter(fragmentActivity: FragmentActivity) :
        FragmentStateAdapter(fragmentActivity) {
        override fun createFragment(position: Int): Fragment {
            return if (position == 0) {
                listCitiesFragment
            } else {
                weatherDetailsFragment
            }
        }

        override fun getItemCount(): Int {
            return 2
        }
    }
}